Lib/concurrent/futures/_base.py

```diff
@@ -582,21 +585,41 @@ def map(self, fn, *iterables, timeout=None, chunksize=1):
         """
         if timeout is not None:
             end_time = timeout + time.monotonic()
+        if prefetch is None:
+            prefetch = self._max_workers
+        if prefetch < 0:
+            raise ValueError("prefetch count may not be negative")
 
-        fs = [self.submit(fn, *args) for args in zip(*iterables)]
+        argsiter = zip(*iterables)
+        initialargs = itertools.islice(argsiter, self._max_workers + prefetch)
+
+        fs = collections.deque(self.submit(fn, *args) for args in initialargs)
 
         # Yield must be hidden in closure so that the futures are submitted
         # before the first iterator value is required.
         def result_iterator():
+            nonlocal argsiter
             try:
-                # reverse to keep finishing order
-                fs.reverse()
                 while fs:
-                    # Careful not to keep a reference to the popped future
                     if timeout is None:
-                        yield fs.pop().result()
+                        res = [fs[0].result()]
                     else:
-                        yield fs.pop().result(end_time - time.monotonic())
+                        res = [fs[0].result(end_time - time.monotonic())]
+
+                    # Got a result, future needn't be cancelled
+                    del fs[0]
+
+                    # Dispatch next task before yielding to keep
+                    # pipeline full
+                    if argsiter:
+                        try:
+                            args = next(argsiter)
+                        except StopIteration:
+                            argsiter = None
+                        else:
+                            fs.append(self.submit(fn, *args))
```

Comment:
If the executor has been shut down, this will raise:
`cannot schedule new futures after shutdown`
But, also the base executor holds no state, so at this level it'll be pretty hard to tell if the executor has been shut down or not.

Reply:
@rdarder 
this could catch that RuntimeError and just return, or the executor could keep a weakset of result_iterators and close them on shutdown